RFC-0017-PyTorch-Operator-Versioning.md

```diff
@@ -0,0 +1,223 @@
+* If the breakage is hard (there is no upgrader), throw “version not supported” error and suggest to regenerate the model.
+
+
+### FC updates
+
+* Except for version bump, no FC update is needed for
+    * The BC/FC break op is not in the model (using [Dynamic versioning](https://github.com/pytorch/pytorch/pull/40279/files))
+    * Server with 2-week FC window
+    * Mobile delivering systems use [compatibility analysis](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/mobile/model_compatibility.cpp) to guard FC break models to be delivered to clients. 
+    * For internal refactors that an operator may call different transient operators, re-tracing is required for mobile to make sure the inputs of tracing based selective build don't have missing operators. Currently, it's guarded by a CI lint and a command line to retrace.
+      
+If needed as a temporary hotfix, an optional downgrader can be used in the backport function shown in the diagram below. It's not planed in this RFC and the discussion is left in the "Open Questions" session.
+
+
+## How does it handle BC and FC?
+
+Aligned with the “Compatibility scenarios” in the [doc of versioning in general](https://docs.google.com/document/d/1nyXmss2O003ZgKrhDmd-kyLNjjMqEXww_2skOXqkks4/edit), we do BC and FC validation at both deploy time and runtime.
+
+### BC
+
+Deploying a new runtime that needs to execute an existing on-device model. 
+
+
+
+* Deploy time
+    * The upgraders must be delivered together with the new runtime. 
+* Runtime
+    * The new runtime load the upgraders lazily: only loads the corresponding upgraders when the version matches. 
+    * The new runtime should always be able to run the old model, unless the old model is “retired”. Error out when the runtime’s op min version > the model's op version.
+
+
+### FC
+
+Deploying a new model to an existing runtime. 
+
+
+
+* Deploy time. Assuming we **keep the op version table of old runtime [query runtime]**,
+    * For each operator in the “new” model, the version number is in the range of version in the “old” runtime. 
+    * Otherwise, if possible, backport the model with old operator schema
+* Runtime
+    * The “old” runtime can run the “new” model, and errors out at load time:
+        * When an unknown op appears.
+        * When reaching an op whose minimum runtime version >= current
+    
+# Open Questions
+
+## Use deprecation window to handle backward compatibility
+One future option is to keep both old and new operators, but set a certain deprecation window for old operators. Deprecate the old operator when the window expires. There are some open questions on this option:
+* What would be the window length? Would it be different for different situations (internal vs. external, server vs. mobile, etc.)
+* From user's point of view, the number of operators my bloat, depending on the length of deprecation window. 
```

Comment:
We have to deprecate for at least 90 days from a user's perspective, anyway, because of our OSS promise